library/core/qsl-base/src/main/java/org/quiltmc/qsl/api/event/Event.java

```diff
@@ -0,0 +1,118 @@
+ *
+ * public void registerEvents() {
+ *     // Since runnable is a functional interface, we can use the lambda form.
+ *     EXAMPLE.register(() -> {
+ *         // Do something
+ *     });
+ *
+ *     // Or we can use a method reference.
+ *     EXAMPLE.register(this::runSomething);
+ *
+ *     // Or implement the callback onto a class.
+ *     // You can also use an anonymous class here; for brevity that is not included.
+ *     EXAMPLE.register(new ImplementedOntoClass());
+ * }
+ *
+ * public void runSomething() {
+ *     // Do something else
+ * }
+ *
+ * // When implementing onto a class, the class must implement the same type as the event invoker.
+ * class ImplementedOntoClass implements Runnable {
+ *     public void run() {
+ *         // Do something else again
+ *     }
+ * }
+ * </code></pre>
+ *
+ * <h2>Example: Executing an event</h2>
+ *
+ * Executing an event is done by calling a method on the event invoker. Where {@code T} is Runnable, executing an event
+ * is done through the following:
+ *
+ * <pre><code>
+ * EXAMPLE.invoker().run();
+ * </code></pre>
+ *
+ * <h2>Best practices for implementations of {@link Event}</h2>
+ *
+ * Implementations do not need to follow all the best practices, but you should consider providing the following:
+ * <ul>
+ *     <li>Synchronization when registering callbacks.
+ *     <ul>
+ *         <li>Registration of callbacks may occur on multiple threads, so thread safety when registering callbacks may
+ *         be useful.
+ *         <li>When the invoker is being executed, a thread safe implementation should ensure the current invoker's
+ *         state should not change. This means a newly registered invoker should only be visible for execution after
+ *         any current open executions of the invoker finish.
+ *     </ul>
+ * </ul>
+ *
+ * @param <T> the type of the invoker used to execute an event
+ */
+public abstract class Event<T> {
+	/**
+	 * The invoker field.
+	 *
+	 * <p>This field is volatile to allow implementations of this class to use a suitable thread safety mechanism if
+	 * desired.
+	 */
+	private volatile T invoker;
+
+	/**
+	 * Returns the invoker instance.
+	 *
+	 * <p>This method is used to invoke an event and execute hidden callbacks implemented by the invoker.
+	 *
+	 * @return the invoker instance
+	 */
+	public final T invoker() {
+		return this.invoker;
+	}
+
+	/**
+	 * Sets the invoker instance.
+	 *
+	 * <p>This method is protected and final since it is only intended for implementors of the {@link Event} class to
+	 * set the invoker instance.
+	 *
+	 * <p>Since the invoker field is volatile, the invoker field should be available immediately for use.
+	 *
+	 * @param invoker the invoker
+	 */
+	protected final void setInvoker(T invoker) {
+		this.invoker = invoker;
+	}
+
+	/**
+	 * Register a callback to the event.
+	 *
+	 * @param callback the callback.
+	 */
+	public abstract void register(T callback);
```

Comment:
If we wanted to add callback IDs and dependencies, now would be the time to do it. The signature of `register` would then be something like
```java
/**
	 * Register a callback to the event.
	 *
     * @param id the identifier of this listener
	 * @param callback the callback
	 * @param after an array of identifiers this callback needs to be run after, if they are present
	 */
public abstract void register(Identifier id, T callback, Identifier... after
```
This can fairly easily be implemented in a way that does not impact invocation-time performance. It does not solve all possible problems that might arise from invoker order, but it does offer one more (relatively pain free) mechanism to mod developers.

It's worth noting that for the purpose of Fabric mod compatibility, any implementation needs to be capable of handling cases where `id` is null, invoking it at any time.

Reply:
`before` is necessary as well.

Reply:
> `before` is necessary as well.

That complicates things slightly as we can no longer just use varargs, but would be fine with some overrides (and probably better method names 😛).
```java
/**
 * Register a callback to the event.
 *
 * @param id the identifier of this listener
 * @param callback the callback
 */
public void register(Identifier id, T callback) {
  this.register(id, callback, null, null);
}

/**
 * Register a callback to the event.
 *
 * @param id the identifier of this listener
 * @param callback the callback
 * @param before an array of identifiers this callback needs to be run before, if they are present
 */
public void registerBefore(Identifier id, T callback, T... before) {
  this.register(id, callback, before, null);
}

/**
 * Register a callback to the event.
 *
 * @param id the identifier of this listener
 * @param callback the callback
 * @param after an array of identifiers this callback needs to be run after, if they are present
 */
public void registerAfter(Identifier id, T callback, T... after) {
  this.register(id, callback, null, after);
}

/**
 * Register a callback to the event.
 *
 * @param id the identifier of this listener
 * @param callback the callback
 * @param before an array of identifiers this callback needs to be run before, if they are present
 * @param after an array of identifiers this callback needs to be run after, if they are present
 */
public abstract void register(Identifier id, T callback, T[] before, T[] after);
```

Reply:
This seems very workable!

Though, shouldn't those `T[]`/`T...` arrays be `Identifier[]`/`Identifier...`, according to the javadocs?

Reply:
> 
> 
> This seems very workable!
> 
> Though, shouldn't those `T[]`/`T...` arrays be `Identifier[]`/`Identifier...`, according to the javadocs?

Oops, yes, I knew I missed something.

Reply:
Varargs cannot be arbitrarily placed, only one per method and it must be last argument

Reply:
I know that, I was pointing out the incorrect typing on the varargs/arrays.

Reply:
Not fully convinced on this. Do not the API is NOT final yet

Reply:
Screenshot of a compromise I might accept for int priorities given some surrounding conditions:

https://media.discordapp.net/attachments/522124386128494599/865243382997844008/Screenshot_20210715-094323_Discord.jpg

Haven seems to be on the other side of the coin on identifiers, I do note that identifiers have some advantages but some disadvantages that integers handle properly

> changing priority ordering is an intentional act, you must have something to be ordered and know what you are ordered before.
>
> no reordering is done given the absence of dependencies

Reply:
There is now an issue for this chain: https://github.com/QuiltMC/quilt-standard-libraries/issues/14